pkgs/build-support/trivial-builders/default.nix

```diff
@@ -152,19 +152,30 @@ rec {
     , meta ? { }
     , allowSubstitutes ? false
     , preferLocalBuild ? true
-    }:
+    , ...
+    }@args:
```

Comment:
`runCommandWith` already supports the `derivationArgs` attribute with the same purpose (which I also think is cleaner). The same interface can be adopted for `writeTextFile` and `writeShellApplication`.

Reply:
`writeTextFile` passes all of its arguments to `runCommandWith` as `derivationArgs`. I don't think this makes sense unless you also support renaming `writeTextFile`'s `checkPhase` argument to `derivationArgs.checkPhase`.

In the sense that `writeShellApplication` wraps `writeTextFile` and `writeTextFile` wraps `runCommand`, this is just extending the existing behavior to forward arbitrary arguments in addition to the explicitly blessed ones.

Reply:
The `...` pattern where you pick certain arguments and forward the rest to some `mkDerivation`-derivative is just a bit too full of holes imo:
- It makes typos not trigger an error, since `mkDerivation` happily accepts pretty much any value and uses them as environment variables.
- It makes it so that you can't use new `mkDerivation` attributes in case they happen to conflict with the picked arguments (what if `mkDerivation` gets a `destination` attribute with different semantics?).
  - And similarly, if somebody passes some derivation attribute, but later that becomes a picked-out function argument, stuff might break.
- It makes the distinction between the stable function interface and internals less clear. In particular setting derivation attributes has the potential to break stuff, whereas the function arguments don't.
- It makes the code less maintainable and harder to understand. Updating function arguments now requires changes in two parts of the code and some thinking. Solutions have [been proposed](https://github.com/NixOS/nix/pull/5590) to this, but I think we should just design better interfaces instead

Reply:
Should I change `writeTextFile`'s interface as well, to be consistent?

Reply:
I think so yeah

Reply:
OK, though I suspect this will cause breakage with callers and confusion around arguments like `meta` and `checkPhase` which are also `stdenv.mkDerivation` arguments.